### PR TITLE
Adding a LoadColumnNameAttribute

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves https://github.com/dotnet/machinelearning/issues/4195

### CODE REVIEW

**Comment by tannergooding** [comment]
CC. @eerhardt, @CESARDELATORRE, @luisquintanilla

File: test/Microsoft.ML.Tests/DatabaseLoaderTests.cs [hunk 1]
```diff
@@ -202,6 +305,19 @@ public class IrisVectorData
             public float[] PetalInfo;
         }
 
+        public class IrisVectorDataWithLoadColumnName
+        {
+            public int Label;
+
+            [LoadColumnName("SepalLength", "SepalWidth")]
```
Comment (CESARDELATORRE) [inline test/Microsoft.ML.Tests/DatabaseLoaderTests.cs]: Not sure I understand this. What if you have hundreds of columns to be loaded as a single vector. Do you need to specify all the column names? Can't you specify a column index range?

Comment (tannergooding) [reply test/Microsoft.ML.Tests/DatabaseLoaderTests.cs]: What would it mean to have a name and range? Why not just use the more efficient index based LoadColumn attribute at that point?

Comment (CESARDELATORRE) [reply test/Microsoft.ML.Tests/DatabaseLoaderTests.cs]: Sorry, that's what I mean, to use the LoadColumn attribute for those cases. I was thinking about the `LoadColumn` attribute. I didn't realize that here you are using `LoadColumnName`, not the same attribute, ok. ;)

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/LoadColumnNameAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        /// Maps member to set of fields in database.
+        /// </summary>
+        /// <param name="fieldNames">Distinct database field names to load as part of this column.</param>
+        public LoadColumnNameAttribute(params string[] fieldNames)
+        {
+            Sources = new List<string>(fieldNames);
+        }
+
+        [BestFriend]
+        internal List<string> Sources;
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/Microsoft.ML.Experimental/DataLoadSave/Database/LoadColumnNameAttribute.cs]: I just followed what `LoadColumnAttribute` was doing here. I'm fine with changing.

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs [hunk 3]
```diff
@@ -490,6 +535,7 @@ public Bindings(ModelLoadContext ctx, DatabaseLoader parent)
                 //     ulong: count for key range
                 //   int: number of segments
                 //   foreach segment:
+                //     int: name
```
Comment (eerhardt) [inline src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs]: `int` => `string`

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs [hunk 4]
```diff
@@ -368,19 +401,23 @@ public static ColInfo Create(string name, PrimitiveDataViewType itemType, Segmen
                 if (segs != null)
                 {
                     var order = Utils.GetIdentityPermutation(segs.Length);
-                    Array.Sort(order, (x, y) => segs[x].Min.CompareTo(segs[y].Min));
 
-                    // Check that the segments are disjoint.
-                    for (int i = 1; i < order.Length; i++)
+                    if (segs[0].Name is null)
```
Comment (eerhardt) [inline src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs]: I don't understand this check. Why is `segs[0]` special here?

Also- what happens if `segs` is non-null, but is empty?

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs [hunk 5]
```diff
@@ -318,17 +340,28 @@ public sealed class Options
...
                 Min = min;
                 Lim = lim;
                 ForceVector = forceVector;
             }
+
+            public Segment(string name, bool forceVector)
+            {
+                Contracts.Assert(name != null);
+                Name = name;
+                Min = 0;
```
Comment (eerhardt) [inline src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs]: Maybe `-1` would be better for these? Since that is an invalid value. Rather than using `0`, which could possibly work.

**Comment by eerhardt** [comment]
>             /// Source index range(s) of the column.

This comment is now a bit out-of-date.

`Source index range(s) of the column.`

`Range` can now be more than just `index` ranges.

---
Refers to: src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs:240 in 90316ec. [](commit_id = 90316ec3518890eeeaedab10f6ed1ad1b2b78853, deletion_comment = False)

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs [hunk 6]
```diff
@@ -287,6 +306,9 @@ public Range(int min, int max)
             [Argument(ArgumentType.AtMostOnce, HelpText = "Last index in the range")]
             public int Max;
 
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Name of the column")]
+            public string Name;
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs]: Maybe `Name` is OK, since the XML class doc for Range says it represents input columns. Maybe if we just update the help text here to be `Name of the input column`?

---
In reply to: [332550925](https://github.com/dotnet/machinelearning/pull/4308#discussion_r332550925) [](ancestors = 332550925)

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoaderCursor.cs [hunk 7]
```diff
@@ -441,9 +465,17 @@ private ValueGetter<VBuffer<double>> CreateVBufferDoubleGetterDelegate(ColInfo c
 
                     foreach (var seg in segs)
                     {
-                        for (int columnIndex = seg.Min; columnIndex < seg.Lim; columnIndex++, i++)
+                        if (seg.Name is null)
+                        {
+                            for (int columnIndex = seg.Min; columnIndex < seg.Lim; columnIndex++, i++)
+                            {
+                                editor.Values[i] = DataReader.IsDBNull(columnIndex) ? default : DataReader.GetDouble(columnIndex);
```
Comment (eerhardt) [inline src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoaderCursor.cs]: Needs to stay `double.NaN`.

Same for the `else` code below.

Maybe we should write a test to ensure we don't break this and `float`.

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs (hunk 3 above)
Comment (tannergooding) [reply src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs]: I think it should be `int: id of name`, to match what it is doing above: https://github.com/dotnet/machinelearning/pull/4308/files/90316ec3518890eeeaedab10f6ed1ad1b2b78853#diff-31c601a2df378d0ae48fc3007c11c5edR531

File: test/Microsoft.ML.Tests/DatabaseLoaderTests.cs [hunk 8]
```diff
@@ -189,6 +277,21 @@ public class IrisData
             public float PetalWidth;
         }
 
+        public class IrisDataWithLoadColumnName
+        {
+            [LoadColumnName("Label")]
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply test/Microsoft.ML.Tests/DatabaseLoaderTests.cs]: Changed.

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/LoadColumnNameAttribute.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/Microsoft.ML.Experimental/DataLoadSave/Database/LoadColumnNameAttribute.cs]: Changed.

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs (hunk 3 above)
Comment (tannergooding) [reply src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs]: I've changed to `string id: name`

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs (hunk 4 above)
Comment (tannergooding) [reply src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs]: We don't support mixing index based and name based segments, so if the first segment has a name, all of them must be names.

As for `non-null`, but empty, is that supported; why wouldn't we just normalize that to `null`?

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs]: Changed.

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoaderCursor.cs (hunk 7 above)
Comment (tannergooding) [reply src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoaderCursor.cs]: Fixed.

Adding tests will require a new database with null column types.

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs (hunk 4 above)
Comment (eerhardt) [reply src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs]: > We don't support mixing index based and name based segments, so if the first segment has a name, all of them must be names.

I think that would be a good comment to stick in here. It isn't immediately obvious to readers.

> As for non-null, but empty, is that supported

Walking backward through the code, I see that Column.Source is a publicly settable array. An external caller could set that to an empty array. In the `Bindings` constructor, if Column.Source is an empty array, it would cause `segs` here to be an empty array.

---
In reply to: [332593763](https://github.com/dotnet/machinelearning/pull/4308#discussion_r332593763) [](ancestors = 332593763)

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs [hunk 9]
```diff
@@ -278,15 +300,30 @@ public Range(int min, int max)
             /// <summary>
             ///  The minimum index of the column, inclusive.
             /// </summary>
+            /// <remarks>
+            /// This is <c>-1</c> if the range represents a column name.
```
Comment (eerhardt) [inline src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs]: Since these fields are publicly settable, this doesn't always hold true. The following code is possible:

```C#
Range r = new Range();
r.Name = "foo";
r.Min = 3;
```

Maybe we should somehow relay in the XML docs that `Name` will override `Min` and `Max`.

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/LoadColumnNameAttribute.cs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        /// Maps member to set of fields in database.
+        /// </summary>
+        /// <param name="fieldNames">Distinct database field names to load as part of this column.</param>
+        public LoadColumnNameAttribute(params string[] fieldNames)
+        {
+            Sources = new List<string>(fieldNames);
+        }
+
+        [BestFriend]
+        internal IReadOnlyList<string> Sources;
```
Comment (eerhardt) [inline src/Microsoft.ML.Experimental/DataLoadSave/Database/LoadColumnNameAttribute.cs]: This is still using an `internal`ly settable field. Can we use a readonly property instead?

Comment (tannergooding) [reply src/Microsoft.ML.Experimental/DataLoadSave/Database/LoadColumnNameAttribute.cs]: Fixed.

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs (hunk 9 above)
Comment (tannergooding) [reply src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs]: Changed to:
```
/// This value is ignored if <see cref="Name" /> is not <c>null</c>.
```

and
```
/// This value, if non-<c>null</c>, overrides <see cref="Min" /> and <see cref="Max" />.
```

File: src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs (hunk 4 above)
Comment (tannergooding) [reply src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs]: Updated to handle an empty segment.
